### PR TITLE
try reading real swaggers 

### PR DESCRIPTION
**Pull request by MikeStall**
reading swagger files from https://apis.guru/. These are public swaggers for public endpoints. 

Hits some bugs in swagger parser.

### CODE REVIEW

File: src/tests/Microsoft.PowerFx.Connectors.Tests/BasicRestTests.cs [hunk 1]
```diff
@@ -137,5 +140,53 @@ public async Task PetStore_MultiServer()
             // Ensure we use HTTPS protocol
             Assert.Equal("https", apiDoc.GetScheme().Substring(0, 5));
         }
+
+        [Fact]
+        public async Task RealTest()
```
Comment (MikeStall) [inline src/tests/Microsoft.PowerFx.Connectors.Tests/BasicRestTests.cs:145]: ad hoc test to excercise calling

File: src/tests/Microsoft.PowerFx.Connectors.Tests/BasicRestTests.cs [hunk 2]
```diff
@@ -137,5 +140,53 @@ public async Task PetStore_MultiServer()
...
+
+            var swaggerUrl = "https://api.math.tools/yaml/math.tools.numbers.openapi.yaml";
+
+            // Other intersting files:
+            // "https://api.apis.guru/v2/specs/weatherbit.io/2.0.0/swagger.json"
+            // "https://www.weatherbit.io/static/swagger.json"
+
+            var doc = await ReadSwaggerFromUrl(swaggerUrl).ConfigureAwait(false);
+
+            using var client = new HttpClient(); // public auth
```
Comment (MikeStall) [inline src/tests/Microsoft.PowerFx.Connectors.Tests/BasicRestTests.cs:157]: Note this goes to a public endpoint directly and **completely skips APIM.**  So no ConnectionId, and just uses a plain HttpClient, not a PowerPlatformHttpClient.

File: src/libraries/Microsoft.PowerFx.Connectors/OpenApiParser.cs [hunk 3]
```diff
@@ -343,12 +359,21 @@ internal static List<ServiceFunction> Parse(string functionNamespace, OpenApiDoc
...
+
+            // $$$ basePath is just "/", but we expect it to be 'server' from the swagger file. 
+            // eg, "https://api.math.tools"
+
+            DPath theNamespace = DPath.Root.Append(new DName(functionNamespace));            
+
+            if (httpClient != null && httpClient is HttpClient hc && hc.BaseAddress == null && openApiDocument.Servers.Any())
+            {
+                hc.BaseAddress = new Uri(openApiDocument.Servers.First().Url);
+            }
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Connectors/OpenApiParser.cs:376]: @LucGenetier  - this seems dangerous. 
It's mutating the incoming httpClient, which could be shared.

File: src/tests/Microsoft.PowerFx.Connectors.Tests/BasicRestTests.cs [hunk 4]
```diff
@@ -137,5 +142,165 @@ public async Task PetStore_MultiServer()
...
+            // "https://www.weatherbit.io/static/swagger.json"
+
+            OpenApiDocument doc = await ReadSwaggerFromUrl(swaggerUrl).ConfigureAwait(false);
+
+            // No BaseAdress specified, we'll use the 1st one found in the swagger file
+            using var client = new HttpClient(); // public auth             
+            var funcs = config.AddService("Math", doc, client);
+
+            var engine = new RecalcEngine(config);
+            var expr = "Math.numberscardinal({number: 1791941})";
```
Comment (MikeStall) [inline src/tests/Microsoft.PowerFx.Connectors.Tests/BasicRestTests.cs:165]: >var expr = "Math.numberscardinal({number: 1791941})"; [](http://example.com/codeflow?start=12&length=53)

Can't access return result:
Math.numberscardinal({number: 1791941}).contents.cardinal

    [0]: {Error 39-48: Name isn't valid. 'contents' isn't recognized.}
    [1]: {Error 48-57: Invalid use of '.'}

File: src/tests/Microsoft.PowerFx.Connectors.Tests/BasicRestTests.cs [hunk 5]
```diff
@@ -137,5 +142,165 @@ public async Task PetStore_MultiServer()
...
+
+            // https://date.nager.at/       
+            var swaggerUrl = "https://date.nager.at/swagger/v3/swagger.json";
+
+            OpenApiDocument doc = await ReadSwaggerFromUrl(swaggerUrl).ConfigureAwait(false);
+            using var client = new HttpClient() { BaseAddress = new Uri("https://date.nager.at") };            
+            var funcs = config.AddService("Holiday", doc, client);
+
+            var engine = new RecalcEngine(config);            
+            var expr = @"Holiday.PublicHolidayPublicHolidaysV3(2023, ""US"")";
```
Comment (MikeStall) [inline src/tests/Microsoft.PowerFx.Connectors.Tests/BasicRestTests.cs:260]: >Holiday.PublicHolidayPublicHolidaysV3(2023, ""US"") [](http://example.com/codeflow?start=25&length=51)

@LucGenetier - 
Trying to use return value gets a runtime exception.  Unlike above, we do have a type, so not sure where it's wrong. 

Try this expression:
First(Holiday.PublicHolidayPublicHolidaysV3(2023, ""US"")).localName

passes check, but fails at runtime.

**Comment by LucGenetier** [comment]
✅ No public API change.

File: src/libraries/Microsoft.PowerFx.Connectors/OpenApiExtensions.cs [hunk 6]
```diff
@@ -271,8 +271,7 @@ public static ConnectorParameterType ToFormulaType(this OpenApiSchema schema, St
 
                     switch (schema.Format)
                     {                        
-                        case "date":
-                            return new ConnectorParameterType(schema, FormulaType.Date);
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Connectors/OpenApiExtensions.cs:275]: >Date [](http://example.com/codeflow?start=82&length=4)

fwiw - This change surprises me.  
Are we confident it's safe?

File: src/libraries/Microsoft.PowerFx.Connectors/OpenApiParser.cs [hunk 7]
```diff
@@ -343,12 +359,17 @@ internal static List<ServiceFunction> Parse(string functionNamespace, OpenApiDoc
...
+            ValidateSupportedOpenApiDocument(openApiDocument, connectorSettings.IgnoreUnknownExtensions);
 
             List<ServiceFunction> functions = new List<ServiceFunction>();
             string basePath = openApiDocument.GetBasePath();
-            DPath theNamespace = DPath.Root.Append(new DName(functionNamespace));
-            connectorSettings ??= new ConnectorSettings();
+            string server = GetServer(openApiDocument, httpClient);
+
+            // $$$ basePath is just "/", but we expect it to be 'server' from the swagger file. 
+            // eg, "https://api.math.tools"
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Connectors/OpenApiParser.cs:370]: can this comment be updated/removed now? I originally added to tag the bug.

File: src/libraries/Microsoft.PowerFx.Connectors/OpenApiParser.cs [hunk 8]
```diff
@@ -459,6 +480,16 @@ internal static List<ServiceFunction> Parse(string functionNamespace, OpenApiDoc
             return functions;
         }
 
+        internal static string GetServer(OpenApiDocument openApiDocument, HttpMessageInvoker httpClient)
+        {
+            if (httpClient != null && httpClient is HttpClient hc && hc.BaseAddress == null && openApiDocument != null && openApiDocument.Servers.Any())
+            {
+                return openApiDocument.Servers.First().Url;
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Connectors/OpenApiParser.cs:487]: >Servers.First().Url [](http://example.com/codeflow?start=39&length=19)

Is First() right policy?
Should we bias to "https"?

**Comment by LucGenetier** [comment]
✅ No public API change.

File: src/libraries/Microsoft.PowerFx.Connectors/OpenApiExtensions.cs (hunk 6 above)
Comment (LucGenetier) [reply src/libraries/Microsoft.PowerFx.Connectors/OpenApiExtensions.cs:275]: This "date" format is uncommon and wasn't tested

**Comment by LucGenetier** [comment]
✅ No public API change.

**Comment by MikeStall** [comment]
Approved - but since I made first commit, I can't officially approve it.

**Review by LucGenetier — DISMISSED** [review]

**Review by LucGenetier — DISMISSED** [review]

**Review by LucGenetier — APPROVED** [review]

**Comment by LucGenetier** [comment]
✅ No public API change.

File: src/tests/Microsoft.PowerFx.Connectors.Tests/LivePublicSwaggerTests.cs [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+            // "https://www.weatherbit.io/static/swagger.json"
+
+            OpenApiDocument doc = await ReadSwaggerFromUrl(swaggerUrl).ConfigureAwait(false);
+
+            // No BaseAdress specified, we'll use the 1st HTTPS one found in the swagger file
+            using var client = new HttpClient(); // public auth             
+            var funcs = config.AddService("Math", doc, client);
+
+            var engine = new RecalcEngine(config);
+            var expr = "Math.numberscardinal({number: 1791941})";
```
Comment (MikeStall) [inline src/tests/Microsoft.PowerFx.Connectors.Tests/LivePublicSwaggerTests.cs:43]: >var expr = "Math.numberscardinal({number: 1791941})"; [](http://example.com/codeflow?start=12&length=53)

try using return value here 

Math.numberscardinal({number: 1791941}).contents

**Pull request merged**